mosaic/mosaic.go

```diff
@@ -0,0 +1,602 @@
+const (
+	Fit     Scale = "all"
+	None    Scale = "none"
+	Stretch Scale = "stretch"
+	Center  Scale = "center"
+)
+
+// Options contains all configurable settings.
+type Mosaic struct {
+	Blocks         []Block
+	colorMode      int     // 0=none, 1=8colors, 2=256colors, 3=truecolor.
+	OutputWidth    int     // Output width.
+	OutputHeight   int     // Output height (0 for auto).
+	ThresholdLevel uint8   // Threshold for considering a pixel as set (0-255).
+	DitherLevel    float64 // Dithering amount (0.0-1.0).
+	UseFgBgOnly    bool    // Use only foreground/background colors (no block symbols).
+	InvertColors   bool    // Invert colors.
+	ScaleMode      Scale   // fit, stretch, center.
+	Symbols        Symbol  // Which symbols to use: "half", "quarter", "all".
+}
+
+// Create Mosaic
+func New() *Mosaic {
+	return &Mosaic{
+		Blocks:         HalfBlocks,
+		colorMode:      3,      // Truecolor.
+		OutputWidth:    80,     // Default width.
+		OutputHeight:   0,      // Auto height.
+		ThresholdLevel: 128,    // Middle threshold.
+		DitherLevel:    0.0,    // Medium dithering.
+		UseFgBgOnly:    false,  // Use block symbols.
+		InvertColors:   false,  // Don't invert.
+		ScaleMode:      None,   // Fit to terminal.
+		Symbols:        "half", // Use half blocks.
+	}
+}
+
+// PixelBlock represents a 2x2 pixel block from the image.
+type pixelBlock struct {
+	Pixels      [2][2]color.Color // 2x2 pixel grid.
+	AvgFg       color.Color       // Average foreground color.
+	AvgBg       color.Color       // Average background color.
+	BestSymbol  rune              // Best matching character.
+	BestFgColor color.Color       // Best foreground color.
+	BestBgColor color.Color       // Best background color.
+}
+
+type shiftable interface {
+	~uint | ~uint16 | ~uint32 | ~uint64
+}
+
+func shift[T shiftable](x T) T {
+	if x > 0xff {
+		x >>= 8
+	}
+	return x
+}
+
+// Set ScaleMode on Mosaic
+func (m *Mosaic) Scale(scale Scale) *Mosaic {
+	m.ScaleMode = scale
+	return m
+}
+
+// Set UseFgBgOnly on Mosaic
+func (m *Mosaic) OnlyForeground(fgOnly bool) *Mosaic {
+	m.UseFgBgOnly = fgOnly
+	return m
+}
+
+// Set DitherLevel on Mosaic
+func (m *Mosaic) Dither(ditherLevel float64) *Mosaic {
+	m.DitherLevel = ditherLevel
+	return m
+}
+
+// Set ThresholdLevel on Mosaic
+func (m *Mosaic) Threshold(threshold uint8) *Mosaic {
+	m.ThresholdLevel = threshold
+	return m
+}
+
+// Set InvertColors on Mosaic
+func (m *Mosaic) WithInvertColors(invertColors bool) *Mosaic {
+	m.InvertColors = invertColors
+	return m
+}
+
+// Set OutputWidth on Mosaic
+func (m *Mosaic) Width(width int) *Mosaic {
+	m.OutputWidth = width
+	return m
+}
+
+// Set OutputHeight on Mosaic
+func (m *Mosaic) Height(height int) *Mosaic {
+	m.OutputHeight = height
+	return m
+}
+
+// Set Symbols on Mosaic
+func (m *Mosaic) SetSymbols(symbol Symbol) *Mosaic {
+	m.Symbols = symbol
+	return m
+}
+
+// Render creates a new renderer with the given options.
+func (m *Mosaic) Render(img image.Image) string {
```

Comment:
We should also have a package `mosaic.Render(img)` that uses default mosaic options `var defaultMosaic Mosaic`